### ISSUE TITLE
Emoji reaction search only checks beginning of name

### ISSUE THREAD

**Issue by chrisbobbe** [zulip/zulip-mobile#3833]
Along with https://github.com/zulip/zulip-mobile/issues/2927, this would really improve the ease of reacting with an emoji.

I just noticed that if I enter the term “smile” in the search box for choosing a message reaction emoji, the results are “smile,” “smile_cat,” “smiley,” “smiley_cat,” and “smiley_hacker.” But I’m actually looking for “slight_smile.” To get that to show up in the results, I have to start typing “slight...” — looks like it’s only searching the available emoji by the beginning of the name.

**Comment by chrisbobbe** [comment]
I expect this will be a small, easy change to the predicate passed in the call to `names.filter` in the exported getFilteredEmojiNames function in src/emoji/data.js. Marking as a good first issue. 🙂 

Hmm! I just noticed that GitHub has the same limitation. Is there a reason we would want to exclude results where the search term matches something other than the beginning of the name?

**Comment by sumukhah** [comment]
Working on the issue

**Comment by chrisbobbe** [comment]
Thanks @sumukhah!

**Comment by chrisbobbe** [comment]
Hmm, as @gnprice points out, there's a nice feature in the web app where if you start typing an emoji's name in the compose box (not the reaction search box!), it brings up a typeahead to search the emoji:

![image](https://user-images.githubusercontent.com/22248748/73016289-6c711c00-3dd2-11ea-87f6-f18535d31120.png)

The logic for that search is more sophisticated than I originally had in mind for this issue, but I think it would be good to implement here. In particular, see `#2`:

1. It searches the entire name, not just the beginning (as the current issue proposes), and
2. It also _reorders_ the search results, so that if they _happen_ to start with the search term, they are listed first.

`#2` can be seen in the screenshot, above: If I've started my search term with "light," I may want "slight smile" to appear, but I'm probably more interested in results that actually do start with "light": "light bulb," "lightning," etc., so they appear first.

To revisit my original example, you'll notice that if you try ":smile" in the web app, `#2` will push "slight smile" so far down the list (because "smile," "smiley," etc., appear first) that it actually doesn't appear. This is only because the typeahead is limited to 5 items, which isn't currently the case on mobile.

**Comment by chrisbobbe** [comment]
(If you're interested, the code for this in the web app is at https://github.com/zulip/zulip/blob/master/static/third/bootstrap-typeahead/typeahead.js#L193, but a simple copy and paste won't be the best approach; you'll want to re-implement this considering the context and code style of the mobile app.)

**Comment by timabbott** [comment]
Hmm, that's Bootstrap typeahead; the logic for prefix-based sorting is actually in our own JS code, in e.g. `static/js/*typeahead*.js`.  We recently did a big refactoring of that JS library code to improve its performance in situations with 10K+ users in an organization, and we're now pretty happy with the implementation's performance at scale properties.  

It's possible that we could refactor some of that logic to be shared code, and that might be a good idea in the long term.  One reason to consider doing something like that now is that @showell just did a big bundle bunch of work on the typeahead code for that performance work, so it's very much in our mental cache.  It might be too complicated, but I think it is worth mentioning that a lot of hours have gone into typeahead sorting logic in the webapp, and we have thousands of lines of tests for that logic, and this is conceptually a pure-logic type thing where some code sharing might be feasible (potentially with some refactoring to make it possible for the mobile app to stub out modules like `pm_conversations` that are used to prefer users one has had PMs with until it has an analog to use).

**Comment by gnprice** [comment]
> It's possible that we could refactor some of that logic to be shared code, and that might be a good idea in the long term. One reason to consider doing something like that now is that @showell just did a big bundle bunch of work on the typeahead code for that performance work, so it's very much in our mental cache.

Interesting, thanks for that tip!

I definitely think typeahead/autocomplete is an area where it'll very likely be fruitful to share code between web and mobile. As you say, there's a lot of pure logic in here, and it encodes a lot of nuanced product choices that would be a lot of work to reimplement one by one in the mobile app. There's also some annoyances people regularly run into on mobile which the webapp has a nice feature that solves, like #2603 and #3710 -- so there's immediate gain to be had in the mobile experience by sharing this code, in addition to the maintainability benefits.

For this issue, I think there's a pretty low-complexity fix we can do in our existing code (as Chris said above: include non-prefix matches, but just rank them after the prefix matches). So I think it's useful to go ahead and do that.

But, independently, I'd love to switch to sharing much of the typeahead code with the webapp, particularly if now is a good time for you and Steve to be thinking about that code and doing some of the refactoring to share it. I'll open a separate issue for that discussion.

**Comment by gnprice** [comment]
Filed zulip/zulip#13728 for the idea of sharing typeahead code between web and mobile.

**Comment by ashwaniYDV** [comment]
I would like to work upon this issue.

**Comment by chrisbobbe** [comment]
Thanks @sumukhah for your work on this! As Greg said at https://github.com/zulip/zulip-mobile/pull/3836#issuecomment-578980883, it looks like we're making much faster progress on sharing code between the web app and the mobile app than we originally thought, so, unless something happens that delays that work, I'm afraid I won't be reviewing your pull request, and I'm sorry about that.

You may like to read through Tim's and Greg's comments, above, to learn more about the choice to share code between the web app and the mobile app. It's a really exciting win for long-term maintainability of both projects, because after it's done, any improvement to the logic in both apps can be made in just one place. 🙂 It can be a lot of effort in the short term; we have to decide what's shareable and what's not, and some adjustments have to be made to get it to work, etc. On the surface, the effort may even seem pointless, if everything's working well in both projects. (Although, in this case, we've definitely found something to be improved!) But in general, being mindful of reducing (or at least not substantially adding to) the future maintenance burden of a project is an important principle of software development. There are a few issues marked with the @zulip/shared label in this repository where this is happening.

@gnprice, do we currently have documentation for the strategy of sharing code? I didn't find any in a quick search.

**Comment by gnprice** [comment]
> do we currently have documentation for the strategy of sharing code? I didn't find any in a quick search.

I don't think we really have something on the overall strategy. Here are some pieces we do have:
* the description of #13728, in which I wrote up the general roadmap for how to migrate a given piece of web frontend code into shared code
* the [README](https://github.com/zulip/zulip/blob/master/static/shared/README.md) of the `@zulip/shared` package's tree, which discusses how it's deployed
* [docs/howto/yarn-link.md](https://github.com/zulip/zulip-mobile/blob/master/docs/howto/yarn-link.md), which describes a key piece of dev workflow for developing changes in `@zulip/shared` together with the mobile app

To date we've moved just one small subsystem to actually being shared, which is typing notifications, so there hasn't been much experience to write up. Definitely we should write more as the typeahead code proceeds.

**Comment by gnprice** [comment]
This got unblocked earlier this year, with https://github.com/zulip/zulip/pull/13739 and some followups shortly after. There's now a [`typeahead.js`](https://github.com/zulip/zulip/blob/ed183fb95fcbb5cf7a4ded5fdfa24a40e1368d1f/static/shared/js/typeahead.js) in the shared code which covers what we need for emoji typeahead.

Here's the next steps, as described at zulip/zulip#13728:

> The roadmap for doing this would be a lot like our initial proof-of-concept for shared code, with `typing_status.js` (in zulip/zulip#13253, zulip/zulip#13315, [zulip/zulip-mobile#3638](https://github.com/zulip/zulip-mobile/pull/3638), [zulip/zulip-mobile#3662](https://github.com/zulip/zulip-mobile/pull/3662)):
> 
> * [done!] Move the relevant code into `static/shared/`. […]
> * On the mobile side, we'll update to a new version of the `@zulip/shared` NPM package (which comes from `static/shared/`, and replace our own implementations with calls to the shared code.
> * The process can be iterative; as we migrate in the mobile app, we may spot further refactors that'll be helpful to make to the shared code.

So we need to bump the version of `@zulip/shared` that we use, and adapt our code to use the shared functions `typeahead.get_emoji_matcher` and `typeahead.sort_emojis` in place of our own implementations of the same functionality.

In the process of doing that, we may see refactors we want to make to the shared code and perhaps the webapp code that uses it, and if so we can do those along the way.

For how to go about doing this:
* See the two bits of docs I mentioned at https://github.com/zulip/zulip-mobile/issues/3833#issuecomment-585960774 above, for the workflow
* See the four PRs (two zulip-mobile, two zulip/zulip) I linked in the quote above for an example (but note they include some work we only had to do once, and some work on the zulip/zulip side that corresponds to what's already been done now for emoji typeahead)
  * One thing we'll definitely want to do in zulip/zulip is to add types for the interface of this shared code; see https://github.com/zulip/zulip/commit/b70b7df22c36b6f3bf92fb99734cb22697b0a3ee for the previous example of that

**Comment by gnprice** [comment]
Marking as priority because, although the user-visible effect of this feature isn't top-priority issue in itself, this is a good pilot for sharing other code with the webapp, and there's active work on the webapp side on that right now.

**Comment by gnprice** [comment]
I've gone and filed a fresh issue #4636 for the task of switching to shared code here (adapting a couple of my comments above) -- I think that will make it easier to see the status and what's needed next.

With that filed, I've retitled this one to the original user-visible symptom it was about, and also dropped the P1 label because the previous comment no longer applies.

**Comment by gnprice** [comment]
Marking as blocked because our plan is to resolve this by doing #4636.

**Comment by gnprice** [comment]
This was indeed resolved as part of #4636.
